### PR TITLE
FOUNDATIONS: PdsService PatientLookupNhsNumberAsync

### PR DESCRIPTION
**Pull request by JParkesAD**
CLOSES [AB#23308](https://dev.azure.com/NELAnalytics/aef6c175-110f-4ba4-82fc-70761f459236/_workitems/edit/23308)

### CODE REVIEW

File: LondonDataServices.IDecide.Core/Services/Foundations/Pds/PdsService.Validations.cs [hunk 1]
```diff
@@ -16,5 +19,44 @@ private static void ValidatePatientLookupIsNotNull(PatientLookup patientLookup)
...
+                if (rule.Condition)
+            {
+                    invalidPdsException.UpsertDataList(
+                        key: parameter,
+                        value: rule.Message);
+                }
+            }
+
+            invalidPdsException.ThrowIfContainsErrors();
+        }
```
Comment (cjdutoit) [inline LondonDataServices.IDecide.Core/Services/Foundations/Pds/PdsService.Validations.cs:60]: indentation here and the lines below

File: LondonDataServices.IDecide.Core.Tests.Unit/Services/Foundations/Pds/PdsServiceTests.Logic.PatientLookupByNhsNumber.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            Hl7.Fhir.Model.Patient outputFhirPatient = CreateRandomPatientWithNhsNumber(inputNhsNumber);
+            Patient expectedPatient = GeneratePatientFromFhirPatient(outputFhirPatient);
+
+            var pdsServiceMock = new Mock<PdsService>(
+                this.pdsBrokerMock.Object,
+                this.loggingBrokerMock.Object)
+            { CallBase = true };
+
+            pdsServiceMock.Setup(service =>
+                service.MapToPatientFromFhirPatient(outputFhirPatient))
```
Comment (cjdutoit) [inline LondonDataServices.IDecide.Core.Tests.Unit/Services/Foundations/Pds/PdsServiceTests.Logic.PatientLookupByNhsNumber.cs:31]: I can't see a logic or validation test for `MapToPatientFromFhirPatient`

```cs
        virtual internal Patient MapToPatientFromFhirPatient(Hl7.Fhir.Model.Patient fhirPatient)
        {

            // I assume all the props is nullable, but I think you need to validate that fhirPatient in not null
            // as well as an additional logic test to ensure your test pass with whitespace characters
            // also, just double check your redation tests to see if it works with whitespace characters
            
            Patient patient = new Patient
            {
                NhsNumber = fhirPatient.Id,
                Address = GetCurrentAddressString(fhirPatient),
                DateOfBirth = DateTimeOffset.Parse(fhirPatient.BirthDate),
                Email = GetCurrentEmail(fhirPatient),
                GivenName = GetFirstName(fhirPatient),
                Surname = GetSurname(fhirPatient),
                PostCode = GetCurrentPostcode(fhirPatient),
                Phone = GetCurrentPhoneNumber(fhirPatient),
                Gender = GetPatientGender(fhirPatient),
                Title = GetPatientTitle(fhirPatient)
            };

            return patient;
        }
```

File: LondonDataServices.IDecide.Core.Tests.Unit/Services/Foundations/Pds/PdsServiceTests.Logic.PatientLookupByNhsNumber.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                    .Returns(expectedPatient);
+
+            this.pdsBrokerMock.Setup(broker =>
+                broker.PatientLookupByNhsNumberAsync(inputNhsNumber))
+                        .ReturnsAsync(outputFhirPatient);
```
Comment (cjdutoit) [inline LondonDataServices.IDecide.Core.Tests.Unit/Services/Foundations/Pds/PdsServiceTests.Logic.PatientLookupByNhsNumber.cs:36]: indentation

File: LondonDataServices.IDecide.Core.Tests.Unit/Services/Foundations/Pds/PdsServiceTests.Logic.PatientLookupByNhsNumber.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            //then
+            actualPatient.Should().BeEquivalentTo(expectedPatient);
+
+            pdsServiceMock.Verify(service =>
+                service.MapToPatientFromFhirPatient(outputFhirPatient),
+                    Times.Once());
+
+            this.pdsBrokerMock.Verify(broker =>
+                broker.PatientLookupByNhsNumberAsync(inputNhsNumber),
+                        Times.Once);
```
Comment (cjdutoit) [inline LondonDataServices.IDecide.Core.Tests.Unit/Services/Foundations/Pds/PdsServiceTests.Logic.PatientLookupByNhsNumber.cs:53]: indentation

**Review by cjdutoit — CHANGES_REQUESTED** [review]

File: LondonDataServices.IDecide.Core/Services/Foundations/Pds/PdsService.Validations.cs (hunk 1 above)
Comment (JParkesAD) [reply LondonDataServices.IDecide.Core/Services/Foundations/Pds/PdsService.Validations.cs]: Fixed

File: LondonDataServices.IDecide.Core.Tests.Unit/Services/Foundations/Pds/PdsServiceTests.Logic.PatientLookupByNhsNumber.cs (hunk 4 above)
Comment (JParkesAD) [reply LondonDataServices.IDecide.Core.Tests.Unit/Services/Foundations/Pds/PdsServiceTests.Logic.PatientLookupByNhsNumber.cs]: Fixed

File: LondonDataServices.IDecide.Core.Tests.Unit/Services/Foundations/Pds/PdsServiceTests.Exceptions.PatientLookupNhsNumber.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+                expectedpdsServiceException);
+
+            pdsBrokerMock.Verify(broker =>
+                broker.PatientLookupByNhsNumberAsync(inputNhsNumber),
+                        Times.Once);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogErrorAsync(It.Is(SameExceptionAs(
+                    expectedpdsServiceException))),
+                        Times.Once);
```
[the remark this answers is not in the document]
Comment (JParkesAD) [reply LondonDataServices.IDecide.Core.Tests.Unit/Services/Foundations/Pds/PdsServiceTests.Exceptions.PatientLookupNhsNumber.cs]: Fixed

File: LondonDataServices.IDecide.Core.Tests.Unit/Services/Foundations/Pds/PdsServiceTests.Validations.PatientLookupByNhsNumber.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+            PdsValidationException actualException =
+                await Assert.ThrowsAsync<PdsValidationException>(patientLookupByNhsNumberAction.AsTask);
+
+            // then
+            actualException.Should().BeEquivalentTo(expectedPdsValidationException);
+
+            pdsBrokerMock.Verify(broker =>
+                broker.PatientLookupByNhsNumberAsync(invalidNhsNumber),
+                        Times.Never);
```
[the remark this answers is not in the document]
Comment (JParkesAD) [reply LondonDataServices.IDecide.Core.Tests.Unit/Services/Foundations/Pds/PdsServiceTests.Validations.PatientLookupByNhsNumber.cs]: Fixed

**Review by cjdutoit — APPROVED** [review]
